Mobile deployment docs overhaul (#4606)
mattermost/docs: source/mobile/build-custom-mobile-apps.rst

Building and Distributing Your Own Custom Mattermost Mobile Apps
================================================================

You can build and distribute custom versions of the Mattermost Mobile App. Choosing this approach means `you've decided not to use the mobile app Mattermost has made available <https://docs.mattermost.com/mobile/use-prebuilt-mobile-apps.html>`__ through public app stores.

This approach is recommended for:

- Organizations that want to customize their Mattermost Mobile Apps.
- Customers using self-hosted Mattermost Team Edition or Enterprise Editions that prefer to host their own push proxy server instead of using one of Mattermost’s hosted versions.
  
Deployment Options
------------------

When you decide to build your own Mattermost Mobile apps, you have multiple ways to deploy it: 

- Submitting your app to `an Enterprise App Store <#using-an-enterprise-app-store>`_.
- Using `an Enterprise Mobile Management (EMM) provider <https://docs.mattermost.com/mobile/deploy-mobile-apps-using-emm-provider.html>`__.

> - Submitting your app to `public app stores <#using-public-app-stores>`_.

Reviewer: For this one, I believe there are some branding guidelines that they need to follow. see https://handbook.mattermost.com/operations/operations/company-processes/publishing/publishing-guidelines/brand-and-visual-design-guidelines#name-usage-guidelines

I see you added this below, but perhaps makes sense to add it here as well?

Reviewer: 0/5

Author: Given that this content is located on the same page as the link to the branding details in our Handbook, it would be best to limit the number of links to that destination page to a single link.

- Using `another distribution method <#using-another-distribution-method>`_.

Developing and Maintaining Your Custom App
------------------------------------------

The Mattermost Mobile App is an open source project. Customizing Mattermost Mobile apps requires a fork of the source code. Your team will be responsible for maintaining that fork, as well as keeping that fork updated with any changes made by Mattermost.

This means that you manage the maintenance of your custom Mattermost Mobile apps, such as rebuilding and incorporating feature and/or security updates. If this isn't done regularly, your applications won't match the functionality of our publicly-available applications, and could be incompatible with future versions of Mattermost Server.

> This process can be complicated and can greatly increase deployment time, not only initially, but whenever the apps need to be updated. We recommend having your development team `review the Mattermost Mobile Apps developer documentation <https://developers.mattermost.com/contribute/mobile/>`__ to ensure they understand the scale and requirements of taking this path. This documentation provides guidance on building, compiling, signing, and customizing Mattermost Mobile apps.

Reviewer: ```suggestion
This process can be complicated and can greatly increase deployment time, not only initially, but whenever the apps need to be updated. We recommend having your development team `review the Mattermost Mobile Apps developer documentation <https://developers.mattermost.com/contribute/mobile/>`__ to ensure they understand the scale and requirements of taking this path. This documentation provides guidance on building, compiling, signing, and white labeling Mattermost Mobile apps.
```

Customizing is too big, it can mean adding code or functionality and that isn't covered by the dev docs

In general, building your own Mobile apps will present some challenges, including:

- Installing the necessary developer tools (such as Nodejs, XCode Developer Tools, Android SDKs, as well as others).
- Obtaining and providing certificates for your custom Mattermost Mobile apps*.

> - Signing your custom Mattermost Mobile apps*.

Author: ```suggestion
- Signing your custom Mattermost Mobile apps *.
```

> - Signing your custom Mattermost Mobile apps *.

Reviewer 2: ```suggestion
- Signing your custom Mattermost Mobile apps*.
```

- Distributing your Mobile app to your users.

> * Mattermost Mobile apps are signed and have certificates associated with Mattermost and public app stores. This means Mattermost’s Mobile App won’t work if you choose to privately host the Mattermost Push Proxy Service (MPNS). You'll need to build your own custom mobile app.

Reviewer: ```suggestion
* Mattermost Mobile apps are signed and have certificates and keys associated with Mattermost and public app stores. This means Mattermost’s Mobile App won’t work if you choose to privately host the Mattermost Push Proxy Service (MPNS). You'll need to build your own custom mobile app.
```

Perhaps add a link to the docs for `build your own custom mobile app`

Author: Given that this content appears on the same page as the "build your own custom mobile app" content, an additional link is likely not necessary here.

Custom Mobile App Branding
~~~~~~~~~~~~~~~~~~~~~~~~~~

Ensure you select a unique app name that helps users distinguish your version from others, such as "<Your Company Name> Chat". See our `Brand and Visual Design Guidelines <https://handbook.mattermost.com/operations/operations/company-processes/publishing/publishing-guidelines/brand-and-visual-design-guidelines#name-usage-guidelines>`__ in our Company Handbook for branding details.

Deploying Your Custom App
-------------------------

You have a number of ways to deploy your custom app for your users to download to their mobile devices.

Using an Enterprise App Store
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This is the most common way for customers to distribute their apps, and it’s our recommended approach. Once your custom app is added to your own Enterprise App Store, your users can download it from the store directly or from an EMM catalog. 

> Alternatively, you can use an EMM provider to push the application to the user’s device, then use the AppConfig standard to enforce a selection of app-specific controls. See our `Manage App Configuration Using AppConfig <https://docs.mattermost.com/mobile/deploy-mobile-apps-using-emm-provider.html#manage-app-configuration-using-appconfig>`__ documentation for details.

Author: Given that this is a new topic on a branch, I suspect that the preview server can't serve it up as expected. Will review all links to ensure they resolve.

Using an EMM Provider
~~~~~~~~~~~~~~~~~~~~~

> See our `Deployment Using an EMM Provider <https://docs.mattermost.com/mobile/deploy-mobile-apps-using-emm-provider.html>`__ documentation to learn more about deploying your custom Mattermost Mobile apps through an EMM provider.

Reviewer: This also applies to the app we publish in the stores.

Author: Yes, agreed. We have this same heading structure in the topic on using the pre-built apps, and that section also points users to the EMM topic for more information.

> See our `Deployment Using an EMM Provider <https://docs.mattermost.com/mobile/deploy-mobile-apps-using-emm-provider.html>`__ documentation to learn more about deploying your custom Mattermost Mobile apps through an EMM provider.

Author: Given that this is a new topic on a branch, I suspect that the preview server can't serve it up as expected. Will review all links to ensure they resolve.

Using Public App Stores
~~~~~~~~~~~~~~~~~~~~~~~

> This is the same process Mattermost uses to make Mattermost Mobile apps available for everyone. It’s a good choice if you're looking to white-label your own Mobile apps to remove Mattermost branding. To submit an app to the public app stores, you need to submit the app to public app stores for review and approval.

Reviewer 2: ```suggestion
This is the same process Mattermost uses to make Mattermost Mobile apps available for everyone. It’s a good choice if you're looking to white-label your own Mobile apps to remove Mattermost branding. Before your app can be listed on the public app stores, you need to submit it to the public app stores for review and approval.
```

As part of the submission process, you need to identify an update strategy that accounts for the release of new versions of Mattermost Mobile apps. 

This update strategy should include:

- Reviewing compatibility requirements.
- Validating mobile app versions connecting to the server.

> - Updating Mattermost Server.

Reviewer 2: Are we referring to the binary or the physical server?

Author: That's an excellent question. @esethna - Can you confirm?

Reviewer: binary

- Updating Mattermost Mobile apps.

> We highly recommend you update your custom Mattermost Mobile apps to incorporate any security or service releases. Prior to distribution, check any compatibility requirements for the Mobile apps and the Mattermost Server. Consult the `Mattermost Mobile App Changelog <https://github.com/mattermost/mattermost-mobile/blob/master/CHANGELOG.md>`__ and the `Mattermost Server Changelog <https://docs.mattermost.com/administration/changelog.html>`__ for details.

Reviewer 2: ```suggestion
We highly recommend you update your custom Mattermost Mobile apps to incorporate any security or service releases. Prior to distribution, check any compatibility requirements for the Mobile apps and the Mattermost server. Consult the `Mattermost Mobile App Changelog <https://github.com/mattermost/mattermost-mobile/blob/master/CHANGELOG.md>`__ and the `Mattermost Server Changelog <https://docs.mattermost.com/administration/changelog.html>`__ for details.
```

> We highly recommend you update your custom Mattermost Mobile apps to incorporate any security or service releases. Prior to distribution, check any compatibility requirements for the Mobile apps and the Mattermost server. Consult the `Mattermost Mobile App Changelog <https://github.com/mattermost/mattermost-mobile/blob/master/CHANGELOG.md>`__ and the `Mattermost Server Changelog <https://docs.mattermost.com/administration/changelog.html>`__ for details.

Reviewer: the changelog has moved https://docs.mattermost.com/administration/mobile-changelog.html

Author: Thanks! Link has been updated accordingly.

> Building and Distributing Your Own Custom Mattermost Mobile Apps

Reviewer 3: We need a plug in this doc to the notifications doc for setting up push notificaitons, similar what we have in the pre-built apps page.

Author: Great call!

In general, building your own Mobile apps will present some challenges, including:

- Installing the necessary developer tools (such as Nodejs, XCode Developer Tools, Android SDKs, as well as others).
- Obtaining and providing certificates for your custom Mattermost Mobile apps*.
- Signing your custom Mattermost Mobile apps*.
- Distributing your Mobile app to your users.

***** Mattermost Mobile apps are signed, and they have certificates and keys associated with Mattermost and public app stores. This means Mattermost’s Mobile App won’t work if you choose to privately host the Mattermost Push Proxy Service (MPNS). You'll need to build your own custom mobile app.

Custom Mobile App Branding
~~~~~~~~~~~~~~~~~~~~~~~~~~

> Ensure you select a unique app name that helps users distinguish your version from others, such as "<Your Company Name> Chat". See our `Brand and Visual Design Guidelines <https://handbook.mattermost.com/operations/operations/company-processes/publishing/publishing-guidelines/brand-and-visual-design-guidelines#name-usage-guidelines.html>`__ in our Company Handbook for details.

Reviewer: > <Your Company Name> Chat

do we want to call it "Chat" or would it be better as "Messaging"?

Author: Messaging is likely a better choice. Thanks!

Deploying Your Custom App
-------------------------

You have a number of ways to deploy your custom app for your users to download to their mobile devices.

Using an Enterprise App Store
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This is the most common way for customers to distribute their apps, and it’s our recommended approach. Once your custom app is added to your own Enterprise App Store, your users can download it from the store directly or from an EMM catalog. 

Using an EMM Provider
~~~~~~~~~~~~~~~~~~~~~

Alternatively, you can use an EMM provider to push the application to the user’s device, then use the AppConfig standard to enforce a selection of app-specific controls. 

For additional details, see the following documentation:

- `Deployment Using an EMM Provider <https://docs.mattermost.com/mobile/deploy-mobile-apps-using-emm-provider.html>`__ to learn more about deploying your custom Mattermost Mobile apps through an EMM provider.
- `Manage App Configuration Using AppConfig <https://docs.mattermost.com/mobile/mobile-appconfig.html>`__  to learn more about managing your app configuration using App Config.

> - `Mattermost AppConfig Values <https://docs.mattermost.com/mobile/mobile-appconfig.html#mattermost-appconfig-values.html>`__  for information on the configuration options that can be sent from the EMM provider to Mattermost Mobile apps. 

Reviewer: this links points to the same document as the one before it, maybe group the descriptions ?

Author: Great feedback. I'll do exactly that. Thanks!

Using Public App Stores
~~~~~~~~~~~~~~~~~~~~~~~

> This is the same process Mattermost uses to make Mattermost Mobile apps available for everyone. It’s a good choice if you're looking to white-label your own Mobile apps to remove Mattermost branding. Before your app can be listed on the public app stores, you need to submit it to the public app stores for review and approval.

Reviewer: > It’s a good choice if you're looking to white-label your own Mobile apps to remove Mattermost branding.

🤔 

I'm really confused by this.. I think it should be about distribution not about white-label or anything else.. this should be about reaching users.

Author: Fair point! Sounds like we want to remove this sentence altogether.
